app/controllers/media_controller.rb

```diff
@@ -0,0 +1,116 @@
+class MediaController < ApplicationController
+  def index
+    @movies = Movie.order('votes DESC').first(10)
+    @books = Book.order('votes DESC').first(10)
+    @albums = Album.order('votes DESC').first(10)
+  end
+
+  def all
+    medium_type = params[:type]
+    model_class = type_to_model(medium_type)
+    @media = model_class.order('votes DESC')
+    @medium_type_name = medium_type.capitalize
+
+  end
+
+  def show
+    @medium_id = params[:id].to_i
+    @medium = Medium.find_by_id(@medium_id)
+
+    if @medium == nil
+      return render :file => "#{Rails.root}/public/404.html", :status => 404
+    end
+
+    @type = @medium.type.downcase.to_sym
+    return render "show"
+  end
+
+
+  def new
+    @medium_type = params[:type]
+    model_class = type_to_model(@medium_type)
+    @medium = model_class.new
+  end
+
+  def create
+    @medium_type = params[:type]
+    model_class = type_to_model(@medium_type)
+    @medium = model_class.new
+    medium_hash = params[@medium_type.downcase.to_sym]
+    @medium.name = medium_hash[:name]
+    @medium.description = medium_hash[:description]
+
+    case @medium_type.downcase
+    when 'book'
+      @medium.author = medium_hash[:author]
+    when 'movie'
+      @medium.director = medium_hash[:director]
+    when 'album'
+      @medium.singer = medium_hash[:singer]
+      @medium.artist = medium_hash[:artist]
+    end
+
+    @medium.save
+
+    return redirect_to action: 'show', id: @medium.id
+    # else
+    #   return render :status => 400
+    # end
+  end
+
+  def edit
+    @medium = Medium.find(params[:id].to_i)
+    @type = @medium.type.downcase
+    return render 'edit'
+  end
+
+  def update
+    @medium = Medium.find(params[:id].to_i)
+
+    medium_hash = params[@medium.type.downcase.to_sym]
+    @medium.name = medium_hash[:name]
+    @medium.description = medium_hash[:description]
+
+    case @medium.type.downcase
+    when 'book'
+      @medium.author = medium_hash[:author]
+    when 'movie'
+      @medium.director = medium_hash[:director]
+    when 'album'
+      @medium.singer = medium_hash[:singer]
+      @medium.artist = medium_hash[:artist]
+    end
```

Comment:
We should refactor this and the `create` action to pull the shared code into a separate method. I think something like this could work:
```ruby
def update_medium(medium, attributes)
  common_attributes = [:name, :description]
  custom_attributes = {
    'book' => [:author],
    'movie' => [:director],
    'album' => [:singer, :artist]
  }

  medium_attributes = common_attributes + custom_attributes[medium.type.downcase]
  medium.update(attributes.select do |attr_name, value|
    medium_attributes.include? attr_name
  end)
end
```